contrib/clients/xrn/xrn.man
.TH XRN 1 "28 November 1988" "X UCB"
.\" $Header: /users/ricks/xrn-11/R4/RCS/xrn.man,v 1.16 89/11/17 12:59:25 ricks Rel $
.\"
.ds XR \fIxrn\fP
.ds NS `.newsrc'
.ds NG \fBNewsgroup\fP
.ds AR \fBArticle\fP
.ds AL \fBAll\fP
.ds AD \fBAdd\fP
.ds XD `~/.Xdefaults'
.\"
.SH NAME
xrn \- an X-based interface to the USENET news system that uses the NNTP remote
news server
.\"
.SH SYNOPSIS
\*(XR is an X-based interface to the USENET news system
that uses the NNTP remote news server for accessing newsgroups
and articles.
By using the NNTP server, users can read news from personal workstations
with the program accessing the news groups and articles from a central
repository on the local area network.
This manual page applies to version 6.3.
.\"
.SH DESCRIPTION
\*(XR
[-breakLength len]
[-lineLength len]
[+/-confirm list] 
[-deadLetters file] 
.br
[-editorCommand command]
[+/-includeHeader]
[+/-includeSep]
[-iconGeometry +X+Y] 
.br
[-leaveHeaders list]
[-stripHeaders list]
[+/-killFiles] 
[-nntpServer name] 
.br
[-minLines number]
[-maxLines number] 
[-newsrcFile file] 
[-printCommand command]
.br
[-organization organization]
[-replyTo name]
[-saveNewsrcFile file] 
[-saveDir directory]
.br
[-signatureFile file] 
[-saveMode mode] 
[-savePostings file] 
[+/-subjectRead] 
.br
[-topLines number]
[-mailer mailer] 
[-tmpDir directory]
[+/-updateNewsrc]
.br
[+/-sortedSubjects]
[-userBitmap bitmap-file]
.br
[-addButtonList list]
[-ngButtonList list]
[-artButtonList list]
.br
[-artSpecButtonList list]
[-allButtonList list]
.br
[-addBindings bindings]
[-ngBindings bindings]
[-artBindings bindings]
[-allBindings bindings]
.\"

Along with the standard toolkit options:
display, geometry, xrm, and iconic.
.\"
.SH BASIC OPERATION
.PP
\fBDon't let the size of this manual page alarm you.\fP
\*(XR is easy to learn on-line without reading
the documentation.
This manual page describes many features that may be obvious to the
casual observer, along with a large section on customization
(such as choosing colors and fonts).
It also describes how to use scroll bars, buttons, and select text;
if you have used an X toolkit application before, the section
titled "BUTTONS, SCROLL BARS, and SELECTION" can be skipped.
.PP
\*(XR uses the \*(NS file to determine what groups need
to be read.
If the \*(NS file does not exist, it is created, and the
user is subscribed to the news group `news.announce.newusers'.
.PP
\*(XR has four modes of operation: \*(AD, \*(NG, \*(AL,
and \*(AR modes.
\*(AD mode will be entered on startup if there are any groups
that the news system knows about that are not in the \*(NS file
(\fIi.e.,\fP new groups).
In \*(AD mode, the user is given a
list of new groups.
Groups can then be subscribed to and placed in the \*(NS file
at the first position, the last position, or after a group already
in the \*(NS file.
When \*(AD mode is exited, any remaining groups are
added unsubscribed, so the user is not asked about them the
next time \*(XR is started.
On exit from \*(AD mode, or on startup if there are no new groups,
\*(NG mode is entered.
\*(NG mode displays the subscribed to groups that have unread
articles and the range of available articles.
The basic functions available in this mode allow the user to
read a group,
mark all articles in a group as read, 
unsubscribe from a group,
move the cursor around the newsgroup window,
change the order of the list of newsgroups,
re-visit the most recently visited group,
and quit \*(XR.
In addition, the user can
subscribe to a group and specify its position in the \*(NS file, 
query the news server for new articles and groups, and
go to groups that are either not subscribed to or currently have
no unread articles (\fIi.e.,\fP groups not displayed on the screen).
From \*(NG mode the user can go into \*(AL mode.
In \*(AL mode the user is presented with a sorted list of
all known groups and their subscription status (subscribed
or unsubscribed) and can change their status or location
in the \*(NS file.
On exiting \*(AL mode the user is placed back in \*(NG mode.
In order to read the articles in a particular group, the user goes
from \*(NG mode to \*(AR mode.
In \*(AR mode the user can sequence through the articles in
the group forward or backward, mark a group of articles
as read or unread, mark all articles in the current group
as read, unsubscribe to the current group, 
return to the last article visited, search forward or backward
for an article subject (either for the exact subject or for
a regular expression in the subject), locally kill all articles with a
particular subject, and quit (saving all changes) or exit
(leaving all articles marked unread).
In addition, the user can save the current
article in a file, post an article to the group, post a followup
to the current article,
mail a reply to the author of an article, and
return to \*(NG mode.
.\"
.SH NEWS SYSTEM
.sp
The news system is a set of bulletins, discussion groups, program sources,
and other bits of information distributed around the world under the
name `USENET'.
The information is generally called `news' and is broken up into
`newsgroups'.
Each newsgroup deals with a subject or set of subjects.
The subjects for newsgroups are varied:  from discussions about particular
versions of UNIX to movie reviews, from information on the X window system
to commentary on current social and political issues.
.PP
For information on what newsgroups are available, answers to commonly
asked questions, and newsgroup ediquette, read the articles in the
newsgroup `news.announce.newsusers'.
.\"
.SH NEWS SERVER
In order to run \*(XR on a machine, the machine must be running an NNTP
news server or be on a network
that has a machine running an NNTP news server
(the source for an NNTP server can be retrieved by anonymous ftp
from the machine `ucbvax.berkeley.edu' [128.32.137.3, 10.2.0.78]).
Either the file `/usr/local/lib/rn/server' must contain the name
of the server machine, the environment variable NNTPSERVER must be
set to the name of the server machine, the
\fBnntpServer\fP Xdefault must be set to the name of the server machine,
or the `-nntpServer' flag must be specified on the command line
(with the order of precedence being command line,
Xdefault, environment variable, file).
.\"
.SH SCREEN LAYOUT
The screen displayed by \*(XR consists of seven sections:
a title bar, two scrollable text windows, two information bars,
and two button boxes.
The title bar displays the current version of the program.
The top text window displays information based on the mode.
In \*(AD mode, the window displays all groups that are
not currently in the \*(NS file, one per line.
In \*(NG mode, the
window displays the groups containing unread articles;
each group represented by a line of the form:
.sp
.ce 100
Unread news in <group name>           <num> article(s) [<first>,<last>]
.ce 0
.sp
<group name> is the name of the group,
<num> is the number of unread articles,
and <first> and <last> are the numbers of the first
and last available articles, respectively.
In \*(AR mode, the window displays a list of subjects
for the articles in the current group, with each subject line
being represented by a line of the form:
.sp
.ce 100
[+u][S] <num>    <subject of the article>
.ce 0
.sp
where <num> is the article number, a `+' in the first position
means that the article has been read, a `u' in the first position
means that the article has been marked as unread, and a 'S' in
the second position means that the article has been saved to a file.
The top information bar displays information about the mode, the buttons
in the top button box, and error messages.
The top button box has buttons that are specific to the mode
and apply to the information in the top text window.
The bottom text window displays articles in \*(AR mode
and a list of all known groups and their subscription status in \*(AL
mode.
The bottom information bar displays information about the mode,
the buttons in the bottom button box, and error messages.
The bottom button box has buttons that are specific to the mode
and apply to the information in the bottom text window.

.SH BUTTONS, SCROLL BARS, AND SELECTION
All button and text selection commands are done with the
left mouse button.
Single-line text selection is accomplished by
clicking the left mouse button on the desired line.
Multiple-line selection is accomplished by clicking the
left mouse button on the first line, holding the button down,
dragging the mouse to the last line, and releasing the mouse button.
Multiple-line selection can also be accomplished by left-clicking
on the first line, and right-clicking on the last line.
Selected lines appear in reverse video (the foreground and background
colors are switched).
The text windows are scrolled with the scroll bar on the
left side of the window.
Clicking the left mouse button in the scroll bar will scroll the
text down
some fraction of a page; clicking the right mouse button in 
the scroll bar similarly scrolls up.
Clicking the middle mouse button will scroll over larger areas:
clicking at the top of the scroll bar will
scroll to the top of the text, clicking in the middle will
scroll to the middle of the text, and clicking at the bottom will
scroll to the bottom of the text.
For those who like using the keyboard,
hitting control-V while the mouse cursor is in a scrollable text window
will cause the window to scroll down one page, meta-V will scroll up one
page.
Hitting the space bar (while the mouse cursor is in the top button box)
will do the right thing; scroll the article text window
when appropriate, go to the next article at the end of the current article,
go to newsgroup mode when done with all articles in the current group, and
go to the next group when in newsgroup mode.
Note: for the initial X11R4 release of XRN, scrolling the article text
window by hitting the space key has been disabled.
.\"
.SH MODES
The next few sections describe the individual modes in \*(XR.
Each button in the various modes is described with the label 
for the button and the name of the button.
The name of the button can be used in X resources, in the button
lists, and in the bindings tables. 
.\"
.SH ADD MODE
\*(AD mode is entered when new groups have been detected
(groups that the news system knows about but are not in the \*(NS file).
.IP "quit (addQuit)"
Add remaining groups in the list to \*(NS as unsubscribed;
go to group mode.
.IP "add first (addFirst)"
Add the current group(s) to the beginning of the \*(NS file and
mark as subscribed.
The current group is the selected group(s), or the group
on the line containing the cursor.
.IP "add last (addLast)"
Add the current group(s) to the end of the \*(NS file and
mark as subscribed.
.IP "add after group (addAfter)"
Add the current group(s) after a group already in the \*(NS.
A dialog box is used to allow the user to enter the name of the
group to add the group after.
The mouse cursor must be in the dialog box
for \*(XR to accept text (however, it does not have to be
in the type-in area).
The dialog box has two options: \fBabort\fP and \fBadd\fP.
No other buttons on the screen will work until the user
has selected an option in the dialog box.
Hitting carriage return is the same as clicking the
\fBadd\fP button (in all \*(XR dialog boxes hitting carriage
return is the same as clicking in the rightmost button of the
dialog box).
.IP "add unsubscribed (addUnsub)"
Add the current group(s) to the end of the \*(NS file
and mark as unsubscribed.

.SH NEWSGROUP MODE
\*(NG mode 
informs the user of the groups with unread news and
gives the user control over which groups are visited.
.IP "quit (ngQuit)"
quit \*(XR.
.IP "read group (ngRead)"
Read the articles in the current group.
The current group is either the one selected (if one is selected)
or the one on the line containing the cursor.
If all groups have been read, the user can still access
groups by using the \fBgoto newsgroup\fP command.
Hitting the space bar with the cursor in the top button box will
call this function.
.IP "next (ngNext)"
Move the cursor to the next group, leaving the articles in
the current group untouched.
.IP "prev (ngPrev)"
Move the cursor to the previous group, leaving the articles in
the current group untouched.
.IP "catch up (ngCatchUp)"
Mark all articles in the current group as read.
.IP "subscribe (ngSubscribe)"
Subscribe to a group.
A dialog box is used to allow the user to enter the
name of the group.
The dialog box has the following options:
\fBabort\fP, 
\fBprev group\fP (subscribe to the previous group visited),
\fBfirst\fP (put group in the beginning of the \*(NS file),
\fBlast\fP (put group in the end of the \*(NS file), and
\fBcurrent position\fP (put group at the position of the cursor).
This command can also be used to change the position of a subscribed
group.
Hitting carriage return after typing in the name is the same as
clicking the \fBcurrent position\fP button.
.IP "unsubscribe (ngUnsub)"
Unsubscribe from the current group.
.IP "goto newsgroup (ngGoto)"
Go to a group that is not currently displayed.
The first unread article is displayed.
If all articles in the group have been read, the last
article in the group is displayed.
.IP "all groups (ngAllGroups)"
Display all of the groups that exist, their subscription status,
and a set of buttons for changing the status.
.IP "rescan (ngRescan)"
Query the server for any new groups or articles.
.IP "prev group (ngPrevGroup)"
Re-visit the previous group visited.
.IP "select groups (ngSelect)"
Records the groups currently selected.
If no groups are selected, nothing is recorded.
.IP "move (ngMove)"
Moves the groups previously selected with the
\fBselect groups\fP command to the current
cursor position.
If the cursor is currently inside the groups
to be moved, no groups are moved.
If any other button is invoked after
\fBselect groups\fP and before \fBmove\fP,
the selection is lost, and no groups are
moved.
.IP "exit (ngExit)"
Quit \*(XR, leaving the \*(NS file unchanged since the
last `rescan' operation or `checkpoint'.
If `updateNewsrc' is on, then the \*(NS file is kept up to
date with the last time \*(AR mode was exited.
.IP "checkpoint (ngCheckPoint)"
Update the \*(NS file.  
\*(XR normally updates the \*(NS file on `rescan' and `quit'
in \*(NG mode. 
This allows you to update the \*(NS file without exiting \*(XR
or reconnecting to the server.
.IP "gripe (ngGripe)"
Send a gripe (bug, bug fix, complaint, feature request, etc.) to
the authors of the program.
.IP "post (ngPost)"
Post an article to a newsgroup.  
See "post" under \*(AR mode for more information.
.\"
.SH ALL MODE
\*(AL mode allows the user to change both the subscription status
and the \*(NS file position of any available group.
All operations can apply to multiple groups.
Thus, the user should never have to use a text editor on the \*(NS file.
Hitting the space bar with the cursor in the top button box
will scroll the bottom text window.
.IP "quit (allQuit)"
Update the \*(NS file and return to group mode.
.IP "subscribe (allSub)"
Subscribe to the current group, leaving it
at its current position in the \*(NS file.
.IP "subscribe first (allFirst)"
Subscribe to the current group and add it to
the beginning of the \*(NS file.
.IP "subscribe last (allLast)"
Subscribe to the current group and add it to the end of the \*(NS file.
.IP "subscribe after group (allAfter)"
Subscribe to the current group and add it after a particular group
in the \*(NS file (which is entered with the use of a dialog box).
.IP "unsubscribe (allUnsub)"
Unsubscribe the current group.
.IP "goto group (allGoto)"
Go to the current newsgroup.
.IP "select groups (allSelect)"
.IP "move (allMove)"
Same as the "select groups" and "move" buttons in \*(NG mode.
.IP "toggle order"
Toggle the order of the newsgroups in the window between \*(NS order
and alphabetical order.
.\"
.SH ARTICLE MODE
\*(AR mode is used for reading and manipulating articles
in a single group.
When a group is entered, the list of article subjects
displayed contains those from the first unread
article to the last available article.
Previous articles can be obtained by using the \fBprev\fP or
\fBsubject previous\fP commands.
Hitting the space bar with the cursor in the top button box
in \*(AR mode will ``do the right thing'';
it will scroll an article if there is more of the article to see
and call the \fBnext unread\fP function if there is no more of the article to
see.
.IP "quit (artQuit)"
Update the \*(NS file and return to group mode.
.IP "next unread (artNextUnread)"
Display the selected article if it is unread; otherwise,
display the first unread article AFTER the cursor position.
If no unread articles exist, \*(XR returns to \*(NG mode.
.IP "next (artNext)"
Display the selected article, if any; otherwise,
display the next article in the current group.
Return to \*(NG mode after the last article has been reached.
.IP "prev (artPrev)"
Display the selected article, if any; otherwise,
display the previous article in the current group.
.IP "last (artLast)"
Display the last article accessed before the
currently displayed one.
This command only keeps track of one previously
accessed article, so invoking it repeatedly
simply toggles the display between two articles.
.IP "next newsgroup (artNextGroup)"
Go directly to the next newsgroup with unread news
(bypass newsgroup mode).
.IP "catch up (artCatchUp)"
Mark all articles in the current group as read;
return to \*(NG mode.
It a particular article is selected, catch up from the
beginning of the group to the selected article.
.IP "goto article (artGotoArticle)"
Go to the specified article number.
.IP "mark read (artMarkRead)"
Mark an article (or group of articles) as read.
This command marks either the selected article(s)
or the article the cursor is on, and leaves
the cursor at its current position.
.IP "mark unread (artMarkUnread)"
Mark an article (or group of articles) and unread.
This command marks either the selected article(s)
or the article the cursor is on, and
leaves the cursor at its current position.
When an article is marked as unread, a 'u' is
placed in the far left column next to the article's
subject.
The \fBnext\fP, \fBprev\fP, \fBsubject next\fP,
and \fBsubject prev\fP buttons will all display
this article if they come across it, but the
\fBnext unread\fP button will not.
The only way to mark an article as read once
it has been marked with a 'u' is to use the
\fBmark read\fP function.
.IP "unsubscribe (artUnsub)"
Unsubscribe from the current group; return to \*(NG mode.
.IP "subject next (artSubNext)"
Find and display the next article with the same subject as the current article
(stripping the '[rR][eE]:' garbage).
If there are no more articles with the current subject and
there are more unread articles, the first unread article is
selected.
If there are no more articles with the current subject and there
are no more unread articles, \*(AR mode is exited.
.IP "subject prev (artSubPrev)"
Find and display the previous article with the same subject
as the current article.
When the program must go to the server to get old articles, the 
top information line will display the number of the article being
fetched.
.IP "session kill (artKillSession)"
Mark all articles with the current subject as read, for
this session only.
.IP "local kill (artKillLocal)"
Mark all articles with the current subject as read for this group,
and for this and all future sessions.
.IP "global kill (artKillGlobal)"
Mark all articles with the current subject as read for all groups,
and for this and all future sessions.
.IP "author kill (artKillAuthor)"
Mark all articles by the current author as read for this session only.
.IP "subject search (artSubSearch)"
Begin a regular expression subject search.
When this button is invoked, a window pops up
querying the user for a regular expression (of the
form used in \fBed\fP), and a direction in which to
search.
If a regular expression is not typed in, the
last regular expression is used, and the search direction
is the one specified in the dialog box (this
can be used to switch the direction of the search
without retyping the expression).
.IP "continue (artContinue)"
Continue the last regular expression search
by searching for the same regular expression
in the same direction.
.IP "post (artPost)"
Post an article to the current group.
A scrollable, editable text window will appear with
a header and the user's `~/.signature' file included.
An appropriate header will be generated with a number
of blank header fields that if left blank will be
deleted from the posting.
The editor defaults to the standard Xtoolkit editor
(similar to emacs), and the default can be overridden
by using the command line option -editorCommand or by
setting the same Xdefault (see the section on command
line arguments).
The basic commands include the following
(note that there are no searching commands):
.sp
.in +0.3i
.nf
^F	forward a character
^B	backward a character
^N	next line
^P	previous line
^D	delete next character
^H	delete previous character
^A	start of line
^E	end of line
^K	delete to end of line (if empty, deletes the line)
^W	delete selection
Meta-Y	yank the cut buffer
Meta-<	beginning of file
Meta->	end of file
Meta-I	insert a file (a dialog box appears to accept a filename, no `~' processing)
.fi
.in -0.3i
.sp
The mouse buttons can be used to select text in this window (which can then be
placed in a file, for example).
.sp
There are four buttons at the bottom of the window: \fBabort\fP,
\fBsend\fP (to post the article), \fBsave\fP (to save the article
in a file), and \fBinclude\fP (include the text of the article in
the reply or followup).
Only one post (or followup or reply) window can be active at a time.
.IP "exit (artExit)"
Exit article mode, marking all articles listed
in the top window as unread.
.IP "gripe (artGripe)"
Send a gripe (bug, bug fix, complaint, feature request, etc.) to
the authors of the program.
.IP "save (artSave)"
Save the current article in a file, or use the current article
as standard input to a command.
The save command will pop up a dialog box for the filename.
The dialog box has two buttons: \fBabort\fP and \fBsave\fP.
If the filename begins with a `|', the article will be sent to the command
specified after the `|'.
Otherwise, the article will be appended to the file specified.
If the name is relative (does not begin with `/' or `~'), it will be
prepended by `~/News/'.
If no name is specified, it will be saved in `~/News/<groupname>', where
`<groupname>' is the name of the current group with the first letter
capitalized (follows the \fIrn\fP article saving conventions).
If \-saveMode it set to `subdirs', then `~/News/<groupname>/' will
be used instead of `~/News/'.
.IP "reply (artReply)"
Reply (by mail) to the author of the current article.
See \fBpost\fP for a description of how to create and send a message.
.IP "forward (artForward)"
Send the current article to another person via mail.
.IP "followup (artFollowup)"
Post a followup article to the current article.
See \fBpost\fP for a description of how to create and send an article.
.IP "cancel (artCancel)"
Cancel the current article.
.IP "rot-13 (artRot13)"
Decrypt a encrypted article.
In the newsgroup `rec.humor', occasionally articles are submitted
that may offend certain people or groups of people.
In order to minimize the offense, these articles are posted in
an encrypted form.
This button will decrypt them.
.IP "toggle header (artHeader)"
Show the full text of an article, including the full header.
.IP "print article (artPrint)"
Send the article to the printer (see the `printCommand' command
line option).

.SH CUSTOMIZING XRN
Colors, fonts, and other \*(XR options can be
specified on the command line or using X resources.
With the exception of the display name, all \*(XR options
can be specified using X resources.
Options specified on the command line take precedence over those specified
using X resources.

.SH COMMAND LINE ARGUMENTS
Here are the current command line arguments
(the X resources have the same names and values as
the command line arguments), under VMS all command line
arguments are in lowercase:
.TP 10
.B \-display display
specification of the X display.
.TP 10
.B \-geometry =WxH+X+Y
specification of the \*(XR window size and location.
The window manager may choose to ignore this specification.
.TP 10
.B -iconGeometry +X+Y
specification of the initial \*(XR icon location.
The window manager may choose to ignore this specification.
.TP 10
.B -iconic
start up \*(XR with the window iconified.
.TP 10
.B \-nntpServer hostname
the NNTP server to use.
.TP 10
.B \-newsrcFile file
the newsrc file to use.
Defaults to `~/.newsrc'.
If a file with a name of the form `<newsrcFile>-<nntpServer>'
is found, it will be used.
.TP 10
.B \-saveNewsrcFile file
the saved `.newsrc' filename.
Before the `.newsrc' file is modified on startup,
it is saved in a backup file.
Defaults to `~/.oldnewsrc'.
.TP 10
.B \-saveDir dir
the article saving directory.
Defaults to `~/News' when \-saveMode specifies `onedir', or 
`~/News/newsgroup' when \-saveMode specifies `subdirs'.
.TP 10
.B \-signatureFile file
the signature file to use.
Defaults to `~/.signature'.
.TP 10
.B \-topLines number
the number of lines to be used for the top
text window (where the unread groups and article
subjects are listed).
.TP 10
.B \-saveMode mode
the mode for saving articles; a comma seperated list of options.
The options can be `mailbox' or `normal', `headers' or `noheaders',
and `onedir'  or `subdirs'.
The default is `normal,headers,onedir'.
.TP 10
.B \-leaveHeaders list
the header fields to leave in the article; a comma separated case-insensitive
list of field names (\fIi.e.,\fP subject,from,organization).
This option takes precedence over `stripHeaders'.
.TP 10
.B \-stripHeaders list
the header fields to strip from the article; a comma separated case-insensitive
list of field names (\fIi.e.,\fP keywords,message-id).
.TP 10
.B \-deadLetters file
the name of the file to save failed postings and messages.
Defaults to `~/dead.letters'.
.TP 10
.B \-savePostings file
the name of the file to save postings and messages (via the `save' button
in the composition window).
Defaults to `~/Articles'.
.TP 10
.B \-minLines number
the minimum number of lines above the cursor in the subject line display.
If negative, the subject line display scrolls only at the bottom and
only one line at a time.
.TP 10
.B \-maxLines number
the maximum number of lines above the cursor in the subject line display.
If negative, the subject line scrolls only at the bottom and
only one line at a time.
.TP 10
.B \-mailer mailer
the command to use for mailing replies.
This command must take all of it's input from stardard input
(\*(XR will not build a command line).
The default is `/usr/lib/sendmail -oi -t'.
.TP 10
.B \-tmpDir directory
the directory to use for the temporary storage of articles fetched
from the server.
If this option and the corresponding Xdefault do not exist,
\*(XR will look for the environment variable \fBTMPDIR\fP.
The default is `/tmp'.
.TP 10
.B \+/\-subjectRead
When using the space bar to scroll, when an article
is finished, the space-bar scrolling invokes \fBsubject next\fP
instead of \fBnext unread\fP.
.TP 10
.B \+/\-sortedSubjects
Display the subject lines in the subject window sorted by subject.
.TP 10
.B \+/\-confirm list
Turn on confirmation boxes for the buttons listed.
These boxes pop up to ask the user to verify
the invocation of "dangerous" actions (such as \fBcatch up\fP
and \fBunsubscribe\fP).
The list of buttons is a comma separated list of button names.
The buttons that can be confirmed: \fBngQuit\fP,
\fBngExit\fP, \fBngCatchUp\fP, \fBartCatchUp\fP, \fBngUnsub\fP,
and \fBartUnsub\fP.
.TP 10
.B \+/\-killFiles
turn the use of kill files on/off.
The default is on.
.TP 10
.B \-userBitmap bitmap-file
create the icon from a user specified bitmap file.
If the bitmap file does not exist or has syntax errors,
the default icon will be used.
.TP 10
.B \-editorCommand command

contrib/clients/xrn/xrn.c

#ifndef lint
static char XRNrcsid[] = "$Header: /users/ricks/xrn-11/R4/RCS/xrn.c,v 1.11 89/11/17 12:53:11 ricks Rel Locker: ricks $";
#endif

/*
 * xrn - an X-based NNTP news reader
 *
 * Copyright (c) 1988, 1989, Ellen M. Sentovich and Rick L. Spickelmier.
 *
 * Permission to use, copy, modify, and distribute this software and its
 * documentation for any purpose and without fee is hereby granted, provided
 * that the above copyright notice appear in all copies and that both that
 * copyright notice and this permission notice appear in supporting
 * documentation, and that the name of the University of California not
 * be used in advertising or publicity pertaining to distribution of 
 * the software without specific, written prior permission.  The University
 * of California makes no representations about the suitability of this
 * software for any purpose.  It is provided "as is" without express or
 * implied warranty.
 *
 * THE UNIVERSITY OF CALIFORNIA DISCLAIMS ALL WARRANTIES WITH REGARD TO 
 * THIS SOFTWARE, INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND 
 * FITNESS, IN NO EVENT SHALL THE UNIVERSITY OF CALIFORNIA BE LIABLE FOR
 * ANY SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER
 * RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF
 * CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN 
 * CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

/*
 * xrn.c: set up the main screens
 *
 */

#include "copyright.h"
#include <X11/Xos.h>
#include "utils.h"
#include "config.h"
#include <X11/Intrinsic.h>
#include <X11/Xutil.h>
#include <X11/StringDefs.h>
#include <X11/Xaw/Paned.h>
#include <X11/Xaw/Label.h>
#include <X11/Xaw/Box.h>
#include <X11/Xaw/AsciiText.h>
#include <X11/Xaw/Text.h>
#include "xthelper.h"
#include "xmisc.h"
#include "resources.h"
#include "internals.h"
#include "error_hnds.h"
#include "buttons.h"
#include "xrn.h"


/* global variables that represent the widgets that are dynamically changed */

Widget topLevel;
Widget TopButtonBox;    /* button box containing the command buttons */
Widget BottomButtonBox; /* button box containing the article specific buttons */
Widget TopInfoLine;      /* top button info line                      */
Widget BottomInfoLine;   /* bottom button info line                   */
Widget Text;             /* newsgroup and article subject display     */
Widget ArticleText;      /* article display                           */

int XRNState;            /* XRN status: news and x                    */


/*ARGSUSED*/
main(argc, argv)
int argc;
char **argv;
{
    Widget titlebar, frame;
    static Arg frameArgs[] = {			/* main window description */
	{XtNx, (XtArgVal) 10},
	{XtNy, (XtArgVal) 10},
	{XtNheight, (XtArgVal) 800},
	{XtNwidth, (XtArgVal) 680},
    };
    static char title[LABEL_SIZE];
    static Arg labelArgs[] = {
	{XtNlabel, (XtArgVal) title},
	{XtNskipAdjust, (XtArgVal) True},
    };
    static Arg genericArgs[] = {
	{XtNskipAdjust, (XtArgVal) True},
    };
    static Arg boxArgs[] = {
	{XtNallowResize, (XtArgVal) True},
	{XtNskipAdjust, (XtArgVal) True},
    };
    static XawTextSelectType selarray[] = {XawselectLine, XawselectNull};
    static Arg textArgs[] = {			/* newsgroup/subject text window */
	{XtNstring,  (XtArgVal) ""},
	{XtNtextOptions, (XtArgVal) scrollVertical},
	{XtNselectTypes, (XtArgVal) selarray },
	{XtNskipAdjust, (XtArgVal) False},
    };
    Arg sizeArgs[2], fontArgs[1];
    int fontHeight;
    static Arg articleTextArgs[] = {		/* article/all text window */
	{XtNstring,  (XtArgVal) ""},
	{XtNtextOptions, (XtArgVal) (scrollVertical|wordBreak)},
	{XtNselectTypes, (XtArgVal) selarray},
	{XtNmin, (XtArgVal) 100},
	{XtNmax, (XtArgVal) 800},
	{XtNskipAdjust, (XtArgVal) False},
    };

    XRNState = 0;

    topLevel = Initialize(argc, argv);

    ehInstallSignalHandlers();
    ehInstallErrorHandlers();

    if (app_resources.geometry != NIL(char)) {
	int bmask;
	bmask = XParseGeometry(app_resources.geometry,       /* geometry specification */
			       (int *) &frameArgs[0].value,    /* x      */
			       (int *) &frameArgs[1].value,    /* y      */
			       (int *) &frameArgs[3].value,    /* width  */
			       (int *) &frameArgs[2].value);   /* height */

	/* handle negative x and y values */
	if ((bmask & XNegative) == XNegative) {
	    frameArgs[0].value += (XtArgVal) DisplayWidth(XtDisplay(topLevel),
							  DefaultScreen(XtDisplay(topLevel)));
	    frameArgs[0].value -= (int) frameArgs[3].value;
	}
	if ((bmask & YNegative) == YNegative) {
	    frameArgs[1].value += (XtArgVal) DisplayHeight(XtDisplay(topLevel),
							   DefaultScreen(XtDisplay(topLevel)));
	    frameArgs[1].value -= (int) frameArgs[2].value;
	}
    }
    
    /* create the pane and its widgets */
    
    frame = XtCreateManagedWidget("vpane", panedWidgetClass, topLevel,
				  frameArgs, XtNumber(frameArgs));
    
#ifdef __DATE__
    (void) sprintf(title, "xrn - version %s (compiled on %s)",
		   XRN_VERSION, __DATE__);
#else
    (void) sprintf(title, "xrn - version %s",
		   XRN_VERSION);
#endif
    
    titlebar = XtCreateManagedWidget("titlebar", labelWidgetClass, frame,
				     labelArgs, XtNumber(labelArgs));

    Text = XtCreateManagedWidget("index", asciiStringWidgetClass, frame,
					 textArgs, XtNumber(textArgs));

    XtSetArg(fontArgs[0], XtNheight, 0);
    XtGetValues(Text, fontArgs, XtNumber(fontArgs));

    /* 6 = 2 * yMargin + 2 */
    fontHeight = (int) fontArgs[0].value - 6;
    
    XtSetArg(sizeArgs[0], XtNmin,
	     app_resources.topLines * fontHeight + 4);
    XtSetArg(sizeArgs[1], XtNmax,
	     app_resources.topLines * fontHeight + 4);
    XtSetValues(Text, sizeArgs, XtNumber(sizeArgs));

    TopInfoLine = XtCreateManagedWidget("indexinfo", labelWidgetClass, frame,
					genericArgs, XtNumber(genericArgs));

    TopButtonBox = XtCreateManagedWidget("indexbuttons", boxWidgetClass, frame,
					  boxArgs, XtNumber(boxArgs));

    /* article display text window */
    XtSetArg(articleTextArgs[4], XtNmax, frameArgs[3].value);
    ArticleText = XtCreateManagedWidget("text", asciiStringWidgetClass, frame,
					 articleTextArgs, XtNumber(articleTextArgs));
    
    BottomInfoLine = XtCreateManagedWidget("textinfo", labelWidgetClass, frame,
					   genericArgs, XtNumber(genericArgs));

    BottomButtonBox = XtCreateManagedWidget("textbuttons", boxWidgetClass, frame,
					     boxArgs, XtNumber(boxArgs));

    /* initialize the news system, read the newsrc file */
    initializeNews();
    XRNState |= XRN_NEWS_UP;

    /* set up the text window, mode buttons, and question */
    determineMode();

    /* create the icon */
    xmIconCreate();
    
    XtRealizeWidget(topLevel);

    XtSetArg(fontArgs[0], XtNheight, 0);
    XtGetValues(titlebar, fontArgs, XtNumber(fontArgs));
    XawPanedSetMinMax(titlebar,
		      (int) fontArgs[0].value, (int) fontArgs[0].value);
    
    XtSetArg(fontArgs[0], XtNheight, 0);
    XtGetValues(TopInfoLine, fontArgs, XtNumber(fontArgs));
    XawPanedSetMinMax(TopInfoLine,
		      (int) fontArgs[0].value, (int) fontArgs[0].value);
    
    XtSetArg(fontArgs[0], XtNheight, 0);
    XtGetValues(BottomInfoLine, fontArgs, XtNumber(fontArgs));
    XawPanedSetMinMax(BottomInfoLine,
		      (int) fontArgs[0].value, (int) fontArgs[0].value);
    
    XtSetArg(fontArgs[0], XtNheight, 0);
    XtGetValues(TopButtonBox, fontArgs, XtNumber(fontArgs));
    XtSetArg(fontArgs[0], XtNheight, 0);
    XtGetValues(BottomButtonBox, fontArgs, XtNumber(fontArgs));
    
    /* allow user resizing of the text windows */
    XawPanedSetMinMax(Text, 10, (int) frameArgs[3].value);
    XawPanedSetMinMax(ArticleText, 100, (int) frameArgs[3].value);
    XawPanedAllowResize(Text, True);
    XawPanedAllowResize(ArticleText, True);

    unbusyCursor((Widget) 0, (caddr_t) 0, (caddr_t) 0);
    
    XRNState |= XRN_X_UP;

    XtMainLoop();

    exit(0);
}       

